### PR TITLE
script/run_tox.sh: Don't overwrite the build dir

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/script/run_tox.sh [hunk 1]
```diff
@@ -115,11 +115,20 @@ function main() {
             shift
         fi
         tox_path=$(get_tox_path $test_name)
+        if [ -z "$test_name" ]; then
```
Comment (tchaikov) [inline src/script/run_tox.sh]: i'd suggest refactor the code to guess tox_path and test_name:
```sh
if [ $# -gt 0 ]; then
  test_name=$1
  shift
  tox_path=$(find $source_dir -path "*/$test_name/tox.ini" | xargs -I{} dirname {})
elif [ -e tox.ini ]; then
  tox_path=$(pwd)
else
  echo "--tox-path not specified" >& 2
  exit 2
fi
```

File: src/script/run_tox.sh [hunk 2]
```diff
@@ -115,11 +115,20 @@ function main() {
...
+            test_name=$(basename $tox_path)
+        fi
         venv_path="$build_dir/$test_name"
     else
         test_name=$(basename $tox_path)
     fi
 
+    if [ "$venv_path" -ef "$build_dir" ]; then
```
Comment (tchaikov) [inline src/script/run_tox.sh]: i am not sure if it's a good idea to make `build_dir` a special case. instead i recommend:

```sh
if [ -e "$venv_path" ]; then
  if [ ! -f ${venv_path}/bin/activate ]; then
    echo "venv_path already exists. bailing out" >& 2
    exit 1
  fi
else
  $source_dir/src/tools/setup-virtualenv.sh ${venv_path}
fi
```

**Comment by tchaikov** [comment]
might want to revise the title of commit message to fix the syntax error.

File: src/script/run_tox.sh (hunk 1 above)
Comment (sebastian-philipp) [reply src/script/run_tox.sh]: > i'd suggest refactor the code to guess tox_path and test_name:
> 
> ```shell
> if [ $# -gt 0 ]; then
>   test_name=$1
>   shift
>   tox_path=$(find $source_dir -path "*/$test_name/tox.ini" | xargs -I{} dirname {})
> elif [ -e tox.ini ]; then
>   tox_path=$(pwd)
> else
>   echo "--tox-path not specified" >& 2
>   exit 2
> fi
> ```

If I'm not mistaken, this still doesn't properly set `test_name`, if it is empty and not given, right?

**Comment by sebastian-philipp** [comment]
@tchaikov can you merge this now?

**Comment by sebastian-philipp** [comment]
@tchaikov ping?

File: src/script/run_tox.sh [hunk 3]
```diff
@@ -121,6 +121,10 @@ function main() {
     fi
 
     if [ ! -f ${venv_path}/bin/activate ]; then
+        if [ -d "$venv_path" ]; then
+            echo "$(readlink -f $venv_path) already exist, but is not a virtualenv. test_name empty?"
```
Comment (tchaikov) [inline src/script/run_tox.sh]: syntax error. how is this connected to `test_name`?

Comment (sebastian-philipp) [reply src/script/run_tox.sh]: > how is this connected to test_name

If we're calling this script with `$prog_name --tox-envs py27,py3` and `tox_path` is empty and we're not giving this script a `test_name`, Line 118  will set `        venv_path="$build_dir/$test_name"` to `venv_path == $build_dir`. This is how I destroyed my build dir.

Comment (tchaikov) [reply src/script/run_tox.sh]: > we're not giving this script a `test_name`

i see now.

> Where do you get a syntax error?

s/exist/exists/

also `readlink -f` is not portable. can we just put `$venv_path` ?  see https://www.freebsd.org/cgi/man.cgi?query=readlink .

Comment (sebastian-philipp) [reply src/script/run_tox.sh]: Fixed. Output now looks like so:

```
'/home/sebastian/Repos/ceph/src/script/../../build/' already exists, but is not a virtualenv. test_name empty?
```

Comment (tchaikov) [reply src/script/run_tox.sh]: > '/home/sebastian/Repos/ceph/src/script/../../build/'

hmm, i admit that this is not ideal. probably an alternative is
```sh
cd $venv_path
echo "$PWD already exists, but is not a virtualenv. test_name empty?"
exit 1
```

Comment (sebastian-philipp) [reply src/script/run_tox.sh]: How about

```bash
function realpath() {
    # readlink -f is not portable
    python -c 'import os,sys;print(os.path.realpath(sys.argv[1]))' $1
+}
```
?

**Comment by sebastian-philipp** [comment]
jenkins test docs

**Pull request merged**